psi4/src/psi4/occ/kappa_orb_resp_iter.cc

```diff
@@ -854,29 +509,189 @@ void OCCWave::orb_resp_pcg_uhf() {
+        kappa_newB->scale(alpha);
+        kappa_newB->add(kappaB);
+
+        // Build r-new
+        r_pcg_newA->zero();
+        r_pcg_newA->copy(sigma_pcgA);
+        r_pcg_newA->scale(-alpha);
+        r_pcg_newA->add(r_pcgA);
+        rms_r_pcgA = r_pcg_newA->rms();
+        r_pcg_newB->zero();
+        r_pcg_newB->copy(sigma_pcgB);
+        r_pcg_newB->scale(-alpha);
+        r_pcg_newB->add(r_pcgB);
+        rms_r_pcgB = r_pcg_newB->rms();
+        rms_r_pcg = MAX0(rms_r_pcgA, rms_r_pcgB);
+
+        // Update S
+        S_pcgA->dirprd(Minv_pcgA, r_pcg_newA);
+        S_pcgB->dirprd(Minv_pcgB, r_pcg_newB);
+
+        double delta_old = delta_new;
+        delta_new = r_pcg_newA->dot(S_pcgA) + r_pcg_newB->dot(S_pcgB);
+
+        // Build line search parameter beta
+        if (pcg_beta_type_ == "FLETCHER_REEVES") {
+            beta = delta_new / delta_old;
+        }
+
+        else if (pcg_beta_type_ == "POLAK_RIBIERE") {
+            dr_pcgA->copy(r_pcg_newA);
+            dr_pcgA->subtract(r_pcgA);
+            dr_pcgB->copy(r_pcg_newB);
+            dr_pcgB->subtract(r_pcgB);
+            beta = (S_pcgA->dot(dr_pcgA) + S_pcgB->dot(dr_pcgB)) / delta_old;
+        }
+
+        // Update D
+        D_pcgA->scale(beta);
+        D_pcgA->add(S_pcgA);
+        D_pcgB->scale(beta);
+        D_pcgB->add(S_pcgB);
+
+        // Reset
+        kappaA->zero();
+        r_pcgA->zero();
+        kappaA->copy(kappa_newA);
+        r_pcgA->copy(r_pcg_newA);
+        kappaB->zero();
+        r_pcgB->zero();
+        kappaB->copy(kappa_newB);
+        r_pcgB->copy(r_pcg_newB);
+
+        // RMS kappa
+        rms_kappaA = kappaA->rms();
+        rms_kappaB = kappaB->rms();
+        rms_kappa = MAX0(rms_kappaA, rms_kappaB);
+        rms_pcg = rms_kappa;
+
+        // increment iteration index
+        itr_pcg++;
+
+        // If we exceed maximum number of iteration, break the loop
+        if (itr_pcg >= pcg_maxiter) {
+            pcg_conver = 0;
+            break;
+        }
+
+        if (rms_r_pcg < tol_pcg || rms_kappa < tol_pcg) break;
+
+    } while (std::fabs(rms_pcg) >= tol_pcg);
+
+}  // end orb_resp_pcg_uhf
+
+// Computes the product of the HF MO hessian and the conjugate direction vector, P.
+// P is assumed to be stored in the D PCG dpdfile. Result is written to S.
+void OCCWave::compute_sigma_vector() {
+    dpdbuf4 K;
+    dpdfile2 S, P, F;
+
+    if (reference_ == "RESTRICTED") {
         // Build sigma = A * p
-        // sigma_AI = 4 \sum_{BJ} (AI|BJ) P_BJ
+
+        // sigma_AI = 2 \sum_{B} F_AB P_BJ
         global_dpd_->file2_init(&S, PSIF_OCC_DPD, 0, ID('V'), ID('O'), "Sigma <V|O>");
         global_dpd_->file2_init(&P, PSIF_OCC_DPD, 0, ID('V'), ID('O'), "P <V|O>");
+        // CAUTION! "Fock" libdpd entries are the Fock elements, as you expect, defined in
+        // fock_[alpha/beta]a. "F" is defined in trans_ints_[r/u]hf and is off-diagonal Fock elements.
+        // We need all Fock elements, so we use "Fock", not "F".
```

Comment:
the text caution speaks of "Fock" vs "F", but all code uses of the former are gone and new code is "FD" and "F". just checking that the caution is in sync.